README.md
# Exoskeleton localStorage Adapter v1.1.8

A fork of [Backbone.localStorage](https://github.com/jeromegn/Backbone.localStorage) that eliminates dependencies on jQuery and Underscore. Good when using [Exoskeleton](https://github.com/paulmillr/exoskeleton) as a drop-in replacement for Backbone.

Basically a monkeypatch until [this pull request](https://github.com/jeromegn/Backbone.localStorage/pull/109) is merged.

CHANGELOG.md
# Changelog

## 1.1.8

- Change Backbone.$.Deferred to Backbone.Deferred for Exoskeleton

## 1.1.7 - October 4, 2013

- Fix CommonJS requiring (Brandon Dimcheff)

## 1.1.6 - July 2, 2013

- Don't assume `require` is defined when `exports` is defined (#100)

## 1.1.5 - At some point...

- Can't remember...

## 1.1.4 - May 19, 2013

- #90 Check for localStorage (throw an error if not available)

## 1.1.3 - May 12, 2013

- #79 Added CommonJS support

## 1.1.2 - May 12, 2013

- #82 Upgraded Backbone to version 1.0 in test suite
- Fixed a few bugs after upgrading to 1.0... looks like Backbone is now triggering the events we were previously manually triggering.

## 1.1.1 - May 11, 2013

- #87 Fixes Lodash v1.0.0-rc.1 removal of _#chain
- #86 Fix for when developer has specified in backbone which jQuery version to use by setting Backbone.$

## 1.1.0 - prior to May 11, 2013

- localStorage adapter for Backbone.js

spec/localStorage_spec.js
// Underscore methods:

_extend = function(obj) {
  Array.prototype.slice.call(arguments, 1).forEach(function(source) {
    if (source) {
      for (var prop in source) {
        obj[prop] = source[prop];
      }
    }
  });
  return obj;
};
_isObject = function(obj) {
  return obj === Object(obj);
};
_clone = function(obj) {
  if (!_isObject(obj)) return obj;
  return Array.isArray(obj) ? obj.slice() : _extend({}, obj);
};
_times = function(n, iterator, context) {
  var accum = Array(Math.max(0, n));
  for (var i = 0; i < n; i++) accum[i] = iterator.call(context, i);
  return accum;
};
_values = function(obj) {
  var keys = Object.keys(obj);
  var length = keys.length;
  var values = new Array(length);
  for (var i = 0; i < length; i++) {
    values[i] = obj[keys[i]];
  }
  return values;
};
_identity = function(value) {
  return value;
};
_toArray = function(obj) {
  if (!obj) return [];
  if (Array.isArray(obj)) return Array.prototype.slice.call(obj);
  if (obj.length === +obj.length) return obj.models.map(function(value) {return value});
  return _values(obj);
};



describe("Exoskeleton.localStorage", function(){

  var attributes = {
    string: "String",
    string2: "String 2",
    number: 1337
  };

  describe("on a Collection", function(){

    var Model = Backbone.Model.extend({
      defaults: attributes
    });

    var Collection = Backbone.Collection.extend({
      model: Model,
      localStorage: new Backbone.LocalStorage("collectionStore")
    });

    var collection = new Collection();


    // Clean up before starting
    before(function(){
      collection.localStorage._clear();
    });

    before(function(){
      collection.fetch();
    });

    it("should use `localSync`", function(){
      assert.equal(Backbone.getSyncMethod(collection), Backbone.localSync);
    });

    it("should initially be empty", function(){
      assert.equal(collection.length, 0);
    });


    describe("create", function(){

      var model;

      before(function(){
        model = collection.create({});
      });

      it("should have 1 model", function(){
        assert.equal(collection.length, 1);
      });

      it("should have a populated model", function(){
        var withId = _clone(attributes);
        withId.id = model.id;
        assert.deepEqual(model.toJSON(), withId);
      });

      it("should have assigned an `id` to the model", function(){
        assert.isDefined(model.id);
      });

    });

    describe("get (by `id`)", function(){

      var model;

      before(function(){
        model = collection.create({});
      });

      it("should find the model with its `id`", function(){
        assert.equal(collection.get(model.id), model);
      });

    });

    describe("instances", function(){

      describe("save", function(){

        var model, model2;

        before(function(){
          model = collection.create({});
          model.save({string: "String 0"});
          collection.fetch()
        });

        it("should persist the changes", function(){
          assert.equal(model.get("string"), "String 0");
        });


        describe("with a new `id`", function(){

          before(function(){
            model2 = collection.create({});
            model2.save({id: 1});
            collection.fetch();
          });

          it("should have a new `id`", function(){
            assert.equal(model2.id, 1);
          });

          it("should have kept its old properties", function(){
            var withId = _clone(attributes);
            withId.id = 1;
            assert.deepEqual(model2.toJSON(), withId);
          });

        });


      });

      describe("destroy", function(){

        var beforeFetchLength, afterFetchLength;

        before(function(){
          // Make sure there's at least items in there
          // ... can't rely on previous tests
          _times(5, function(){
            collection.create()
          });
        });

        before(function(){
          var collArray = _toArray(collection)
          collArray.forEach(function(model){
            model.destroy();
          });
          beforeFetchLength = collection.length;
        });

        before(function(){
          collection.fetch();
          afterFetchLength = collection.length;
        });

        it("should have removed all items from the collection", function(){
          assert.equal(beforeFetchLength, 0);
        });

        it("should have removed all items from the store", function(){
          assert.equal(afterFetchLength, 0);
        });

      });

      describe("with a different `idAttribute`", function(){

        var Model2 = Backbone.Model.extend({
          defaults: attributes,
          idAttribute: "_id"
        });

        var Collection2 = Backbone.Collection.extend({
          model: Model2,
          localStorage: new Backbone.LocalStorage("collection2Store")
        });

        var collection2 = new Collection2();

        before(function(){
          collection2.create();
        });

        it("should have used the custom `idAttribute`", function(){
          assert.equal(collection2.models[0].id, collection2.models[0].get("_id"));
        });

      });

    });

  });

  describe("on a Model", function(){

    var Model = Backbone.Model.extend({
      defaults: attributes,
      localStorage: new Backbone.LocalStorage("modelStore")
    });

    var model = new Model();

    before(function(){
      model.localStorage._clear();
    });

    it("should use `localSync`", function(){
      assert.equal(Backbone.getSyncMethod(model), Backbone.localSync);
    });

    describe("fetch", function(){
      it('should fire sync event on fetch', function(done) {
        var model = new Model(attributes);
        model.on('sync', function(){
          done();
        });
        model.fetch();
      });
    });

    describe("save", function(){

      before(function(){
        model.save();
        model.fetch();
      });

      it("should be saved in the store", function(){
        assert.isDefined(model.id);
      });

      describe("with new attributes", function(){

        before(function(){
          model.save({number: 42});
          model.fetch();
        });

        it("should persist the changes", function(){
          assert.deepEqual(model.toJSON(), _extend(_clone(attributes), {id: model.id, number: 42}));
        });

      });

      describe('fires events', function(){
        before(function(){
          this.model = new Model();
        });
        after(function(){
          this.model.destroy();
        });

        it('should fire sync event on save', function(done) {
          this.model.on('sync', function(){
            this.model.off('sync');
            done();
          }, this);
          this.model.save({foo: 'baz'});
        });
      });

    });

    describe("destroy", function(){

      before(function(){
        model.destroy();
      });

      it("should have removed the instance from the store", function(){
        assert.lengthOf(Model.prototype.localStorage.findAll(), 0);
      });

    });

  });

  describe("Error handling", function(){

    var Model = Backbone.Model.extend({
      defaults: attributes,
      localStorage: new Backbone.LocalStorage("modelStore")
    });

    describe("private browsing", function(){

      var model = new Model()
        , oldSetItem = window.localStorage.setItem
        , oldStorageSize = model.localStorage._storageSize
        , error;

      before(function(done){
        model.localStorage._clear();

        // Patch browser conditions for private error.
        model.localStorage._storageSize = function(){ return 0; };
        window.localStorage.setItem = function(){
          var error = new Error();
          error.code = DOMException.QUOTA_EXCEEDED_ERR;
          throw error;
        };

        model.save(attributes, {
          error: function(model, err){
            error = err;
            done();
          }
        })
      });

      it("should return the error in the error callback", function(){
        assert.equal(error, "Private browsing is unsupported");
      });

      it('should throw an error event', function(done){
        model.on('error', function() {
          done();
        });
        model.save();
      });

      after(function(){
        // Unwrap patches.
        model.localStorage._storageSize = oldStorageSize;
        window.localStorage.setItem = oldSetItem;
      })

    });

  });

});

describe("Without Backbone.localStorage", function(){

  describe("on a Collection", function(){
    var Collection = Backbone.Collection.extend()
      , collection = new Collection();

    it("should use `ajaxSync`", function(){
      assert.equal(Backbone.getSyncMethod(collection), Backbone.ajaxSync);
    });
  });

  describe("on a Model", function(){
    var Model = Backbone.Model.extend()
      , model = new Model();

    it("should use `ajaxSync`", function(){
      assert.equal(Backbone.getSyncMethod(model), Backbone.ajaxSync);
    });
  });

});


// For some reason this is not ran when viewed in a browser
// but it is ran when using `mocha-phantomjs`.
describe("AMD", function(){

  require.config({
    paths: {
      backbone: "support/exoskeleton",
      localstorage: "../exoskeleton.localStorage"
    }
  });

  var LocalStorage;

  before(function(done){
    require(["localstorage"], function(LS){
      LocalStorage = LS;
      done()
    });
  });

  it("should be the same as the non-amd usage", function(){
    assert.equal(Backbone.LocalStorage, LocalStorage)
  });

});


spec/localStorage_commonjs_spec.js
describe("Exoskeleton.localStorage in CommonJS environment", function() {

  var LocalStorage = require("../exoskeleton.localStorage");
  var Backbone = require("backbone");

  it("should be the same as the non-CommonJS usage", function(){
    assert.equal(Backbone.LocalStorage, LocalStorage);
  });
});


backbone.localStorage.js
/**
 * Backbone localStorage Adapter
 * Version 1.1.8
 *
 * https://github.com/jeromegn/Backbone.localStorage
 */
(function (root, factory) {
   if (typeof exports === 'object' && typeof require === 'function') {
     module.exports = factory(require("backbone"));
   } else if (typeof define === "function" && define.amd) {
      // AMD. Register as an anonymous module.
      define(["backbone"], function(Backbone) {
        // Use global variables if the locals are undefined.
        return factory(Backbone || root.Backbone);
      });
   } else {
      // RequireJS isn't being used. Assume underscore and backbone are loaded in <script> tags
      factory(Backbone);
   }
}(this, function(Backbone) {
// A simple module to replace `Backbone.sync` with *localStorage*-based
// persistence. Models are given GUIDS, and saved into a JSON object. Simple
// as that.

// Hold reference to Underscore.js and Backbone.js in the closure in order
// to make things work even if they are removed from the global namespace

// Generate four random hex digits.
function S4() {
   return (((1+Math.random())*0x10000)|0).toString(16).substring(1);
};

// Generate a pseudo-GUID by concatenating random hexadecimal.
function guid() {
   return (S4()+S4()+"-"+S4()+"-"+S4()+"-"+S4()+"-"+S4()+S4()+S4());
};

// Our Store is represented by a single JS object in *localStorage*. Create it
// with a meaningful name, like the name you'd give a table.
// window.Store is deprectated, use Backbone.LocalStorage instead
Backbone.LocalStorage = window.Store = function(name) {
  if( !this.localStorage ) {
    throw "Backbone.localStorage: Environment does not support localStorage."
  }
  this.name = name;
  var store = this.localStorage().getItem(this.name);
  this.records = (store && store.split(",")) || [];
};

// Copy used underscore methods here
var ArrayProto = Array.prototype, ObjProto = Object.prototype, FuncProto = Function.prototype;
var slice              = ArrayProto.slice;
var nativeFilter       = ArrayProto.filter;


_extend = function(obj) {
  ps = Array.prototype.slice.call(arguments, 1);
  ps.forEach(function(source) {
    if (source) {
      for (var prop in source) {
        obj[prop] = source[prop];
      }
    }
  });
  return obj;
};

_contains = _include = function(obj, target) {
  if (obj == null) return false;
  if (nativeIndexOf && obj.indexOf === nativeIndexOf) return obj.indexOf(target) != -1;
  return any(obj, function(value) {
    return value === target;
  });
};

_chain = function(obj) {
  return _(obj).chain();
};


_filter = _select = function(obj, iterator, context) {
  var results = [];
  if (obj == null) return results;
  if (nativeFilter && obj.filter === nativeFilter) return obj.filter(iterator, context);
  each(obj, function(value, index, list) {
    if (iterator.call(context, value, index, list)) results.push(value);
  });
  return results;
};

_reject = function(obj, iterator, context) {
  return _filter(obj, function(value, index, list) {
    return !iterator.call(context, value, index, list);
  }, context);
};

//

_extend(Backbone.LocalStorage.prototype, {

  // Save the current state of the **Store** to *localStorage*.
  save: function() {
    this.localStorage().setItem(this.name, this.records.join(","));
  },

  // Add a model, giving it a (hopefully)-unique GUID, if it doesn't already
  // have an id of it's own.
  create: function(model) {
    if (!model.id) {
      model.id = guid();
      model.set(model.idAttribute, model.id);
    }
    this.localStorage().setItem(this.name+"-"+model.id, JSON.stringify(model));
    this.records.push(model.id.toString());
    this.save();
    return this.find(model);
  },

  // Update a model by replacing its copy in `this.data`.
  update: function(model) {
    this.localStorage().setItem(this.name+"-"+model.id, JSON.stringify(model));
    if (!_include(this.records, model.id.toString()))
      this.records.push(model.id.toString()); this.save();
    return this.find(model);
  },

  // Retrieve a model from `this.data` by id.
  find: function(model) {
    return this.jsonData(this.localStorage().getItem(this.name+"-"+model.id));
  },

  // Return the array of all models currently in storage.
  findAll: function() {
    // Lodash removed _#chain in v1.0.0-rc.1
    return this.records.map(function(id){
        return this.jsonData(this.localStorage().getItem(this.name+"-"+id));
      }, this)
  },

  // Delete a model from `this.data`, returning it.
  destroy: function(model) {
    if (model.isNew())
      return false
    this.localStorage().removeItem(this.name+"-"+model.id);
    this.records = _reject(this.records, function(id){
      return id === model.id.toString();
    });
    this.save();
    return model;
  },

  localStorage: function() {
    return localStorage;
  },

  // fix for "illegal access" error on Android when JSON.parse is passed null
  jsonData: function (data) {
      return data && JSON.parse(data);
  },

  // Clear localStorage for specific collection.
  _clear: function() {
    var local = this.localStorage(),
      itemRe = new RegExp("^" + this.name + "-");

    // Remove id-tracking item (e.g., "foo").
    local.removeItem(this.name);

    // Lodash removed _#chain in v1.0.0-rc.1
    // Match all data items (e.g., "foo-ID") and remove.
    var objkeys = Object.keys(local)
    var filtered = _filter(objkeys, function (k) { return itemRe.test(k); })
    filtered.forEach(function (k) { local.removeItem(k); });

    this.records.length = 0;
  },

  // Size of localStorage.
  _storageSize: function() {
    return this.localStorage().length;
  }

});

// localSync delegate to the model or collection's
// *localStorage* property, which should be an instance of `Store`.
// window.Store.sync and Backbone.localSync is deprecated, use Backbone.LocalStorage.sync instead
Backbone.LocalStorage.sync = window.Store.sync = Backbone.localSync = function(method, model, options) {
  var store = model.localStorage || model.collection.localStorage;

  var resp, errorMessage, syncDfd = Backbone.Deferred && Backbone.Deferred(); //If $ is having Deferred - use it.

  try {

    switch (method) {
      case "read":
        resp = model.id != undefined ? store.find(model) : store.findAll();
        break;
      case "create":
        resp = store.create(model);
        break;
      case "update":
        resp = store.update(model);
        break;
      case "delete":
        resp = store.destroy(model);
        break;
    }

  } catch(error) {
    if (error.code === 22 && store._storageSize() === 0)
      errorMessage = "Private browsing is unsupported";
    else
      errorMessage = error.message;
  }

  if (resp) {
    if (options && options.success) {
      if (Backbone.VERSION === "0.9.10") {
        options.success(model, resp, options);
      } else {
        options.success(resp);
      }
    }
    if (syncDfd) {
      syncDfd.resolve(resp);
    }

  } else {
    errorMessage = errorMessage ? errorMessage
                                : "Record Not Found";

    if (options && options.error)
      if (Backbone.VERSION === "0.9.10") {
        options.error(model, errorMessage, options);
      } else {
        options.error(errorMessage);
      }

    if (syncDfd)
      syncDfd.reject(errorMessage);
  }

  // add compatibility with $.ajax
  // always execute callback for success and error
  if (options && options.complete) options.complete(resp);

  return syncDfd && syncDfd.promise();
};

Backbone.ajaxSync = Backbone.sync;

Backbone.getSyncMethod = function(model) {
  if(model.localStorage || (model.collection && model.collection.localStorage)) {
    return Backbone.localSync;
  }

  return Backbone.ajaxSync;
};

// Override 'Backbone.sync' to default to localSync,
// the original 'Backbone.sync' is still available in 'Backbone.ajaxSync'
Backbone.sync = function(method, model, options) {
  return Backbone.getSyncMethod(model).apply(this, [method, model, options]);
};

return Backbone.LocalStorage;
}));

backbone.localStorage-min.js
/**
 * Backbone localStorage Adapter
 * Version 1.1.8
 *
 * https://github.com/jeromegn/Backbone.localStorage
 */(function(a,b){typeof exports=="object"&&typeof require=="function"?module.exports=b(require("backbone")):typeof define=="function"&&define.amd?define(["backbone"],function(c){return b(c||a.Backbone)}):b(Backbone)})(this,function(a){function b(){return((1+Math.random())*65536|0).toString(16).substring(1)}function c(){return b()+b()+"-"+b()+"-"+b()+"-"+b()+"-"+b()+b()+b()}a.LocalStorage=window.Store=function(a){if(!this.localStorage)throw"Backbone.localStorage: Environment does not support localStorage.";this.name=a;var b=this.localStorage().getItem(this.name);this.records=b&&b.split(",")||[]};var d=Array.prototype,e=Object.prototype,f=Function.prototype,g=d.slice,h=d.filter;return _extend=function(a){return ps=Array.prototype.slice.call(arguments,1),ps.forEach(function(b){if(b)for(var c in b)a[c]=b[c]}),a},_contains=_include=function(a,b){return a==null?!1:nativeIndexOf&&a.indexOf===nativeIndexOf?a.indexOf(b)!=-1:any(a,function(a){return a===b})},_chain=function(a){return _(a).chain()},_filter=_select=function(a,b,c){var d=[];return a==null?d:h&&a.filter===h?a.filter(b,c):(each(a,function(a,e,f){b.call(c,a,e,f)&&d.push(a)}),d)},_reject=function(a,b,c){return _filter(a,function(a,d,e){return!b.call(c,a,d,e)},c)},_extend(a.LocalStorage.prototype,{save:function(){this.localStorage().setItem(this.name,this.records.join(","))},create:function(a){return a.id||(a.id=c(),a.set(a.idAttribute,a.id)),this.localStorage().setItem(this.name+"-"+a.id,JSON.stringify(a)),this.records.push(a.id.toString()),this.save(),this.find(a)},update:function(a){return this.localStorage().setItem(this.name+"-"+a.id,JSON.stringify(a)),_include(this.records,a.id.toString())||this.records.push(a.id.toString()),this.save(),this.find(a)},find:function(a){return this.jsonData(this.localStorage().getItem(this.name+"-"+a.id))},findAll:function(){return(_chain||_)(this.records).map(function(a){return this.jsonData(this.localStorage().getItem(this.name+"-"+a))},this).compact().value()},destroy:function(a){return a.isNew()?!1:(this.localStorage().removeItem(this.name+"-"+a.id),this.records=_reject(this.records,function(b){return b===a.id.toString()}),this.save(),a)},localStorage:function(){return localStorage},jsonData:function(a){return a&&JSON.parse(a)},_clear:function(){var a=this.localStorage(),b=new RegExp("^"+this.name+"-");a.removeItem(this.name),(_chain||_)(a).keys().filter(function(a){return b.test(a)}).each(function(b){a.removeItem(b)}),this.records.length=0},_storageSize:function(){return this.localStorage().length}}),a.LocalStorage.sync=window.Store.sync=a.localSync=function(b,c,d){var e=c.localStorage||c.collection.localStorage,f,g,h=a.Deferred&&a.Deferred();try{switch(b){case"read":f=c.id!=undefined?e.find(c):e.findAll();break;case"create":f=e.create(c);break;case"update":f=e.update(c);break;case"delete":f=e.destroy(c)}}catch(i){i.code===22&&e._storageSize()===0?g="Private browsing is unsupported":g=i.message}return f?(d&&d.success&&(a.VERSION==="0.9.10"?d.success(c,f,d):d.success(f)),h&&h.resolve(f)):(g=g?g:"Record Not Found",d&&d.error&&(a.VERSION==="0.9.10"?d.error(c,g,d):d.error(g)),h&&h.reject(g)),d&&d.complete&&d.complete(f),h&&h.promise()},a.ajaxSync=a.sync,a.getSyncMethod=function(b){return b.localStorage||b.collection&&b.collection.localStorage?a.localSync:a.ajaxSync},a.sync=function(b,c,d){return a.getSyncMethod(c).apply(this,[b,c,d])},a.LocalStorage});

exoskeleton.localStorage-min.js
/**
 * Backbone localStorage Adapter
 * Version 1.1.8
 *
 * https://github.com/jeromegn/Backbone.localStorage
 */(function(a,b){typeof exports=="object"&&typeof require=="function"?module.exports=b(require("backbone")):typeof define=="function"&&define.amd?define(["backbone"],function(c){return b(c||a.Backbone)}):b(Backbone)})(this,function(a){function b(){return((1+Math.random())*65536|0).toString(16).substring(1)}function c(){return b()+b()+"-"+b()+"-"+b()+"-"+b()+"-"+b()+b()+b()}a.LocalStorage=window.Store=function(a){if(!this.localStorage)throw"Backbone.localStorage: Environment does not support localStorage.";this.name=a;var b=this.localStorage().getItem(this.name);this.records=b&&b.split(",")||[]};var d=Array.prototype,e=Object.prototype,f=Function.prototype,g=d.indexOf,h=d.slice,i=d.filter;return _extend=function(a){return ps=Array.prototype.slice.call(arguments,1),ps.forEach(function(b){if(b)for(var c in b)a[c]=b[c]}),a},_contains=_include=function(a,b){return a==null?!1:g&&a.indexOf===g?a.indexOf(b)!=-1:any(a,function(a){return a===b})},_chain=function(a){return _(a).chain()},_filter=_select=function(a,b,c){var d=[];return a==null?d:i&&a.filter===i?a.filter(b,c):(each(a,function(a,e,f){b.call(c,a,e,f)&&d.push(a)}),d)},_reject=function(a,b,c){return _filter(a,function(a,d,e){return!b.call(c,a,d,e)},c)},_extend(a.LocalStorage.prototype,{save:function(){this.localStorage().setItem(this.name,this.records.join(","))},create:function(a){return a.id||(a.id=c(),a.set(a.idAttribute,a.id)),this.localStorage().setItem(this.name+"-"+a.id,JSON.stringify(a)),this.records.push(a.id.toString()),this.save(),this.find(a)},update:function(a){return this.localStorage().setItem(this.name+"-"+a.id,JSON.stringify(a)),_include(this.records,a.id.toString())||this.records.push(a.id.toString()),this.save(),this.find(a)},find:function(a){return this.jsonData(this.localStorage().getItem(this.name+"-"+a.id))},findAll:function(){return this.records.map(function(a){return this.jsonData(this.localStorage().getItem(this.name+"-"+a))},this)},destroy:function(a){return a.isNew()?!1:(this.localStorage().removeItem(this.name+"-"+a.id),this.records=_reject(this.records,function(b){return b===a.id.toString()}),this.save(),a)},localStorage:function(){return localStorage},jsonData:function(a){return a&&JSON.parse(a)},_clear:function(){var a=this.localStorage(),b=new RegExp("^"+this.name+"-");a.removeItem(this.name);var c=Object.keys(a),d=_filter(c,function(a){return b.test(a)});d.forEach(function(b){a.removeItem(b)}),this.records.length=0},_storageSize:function(){return this.localStorage().length}}),a.LocalStorage.sync=window.Store.sync=a.localSync=function(b,c,d){var e=c.localStorage||c.collection.localStorage,f,g,h=a.Deferred&&a.Deferred();try{switch(b){case"read":f=c.id!=undefined?e.find(c):e.findAll();break;case"create":f=e.create(c);break;case"update":f=e.update(c);break;case"delete":f=e.destroy(c)}}catch(i){i.code===22&&e._storageSize()===0?g="Private browsing is unsupported":g=i.message}return f?(d&&d.success&&(a.VERSION==="0.9.10"?d.success(c,f,d):d.success(f)),h&&h.resolve(f)):(g=g?g:"Record Not Found",d&&d.error&&(a.VERSION==="0.9.10"?d.error(c,g,d):d.error(g)),h&&h.reject(g)),d&&d.complete&&d.complete(f),h&&h.promise()},a.ajaxSync=a.sync,a.getSyncMethod=function(b){return b.localStorage||b.collection&&b.collection.localStorage?a.localSync:a.ajaxSync},a.sync=function(b,c,d){return a.getSyncMethod(c).apply(this,[b,c,d])},a.LocalStorage});

exoskeleton.localStorage.js
/**
 * Backbone localStorage Adapter
 * Version 1.1.8
 *
 * https://github.com/jeromegn/Backbone.localStorage
 */
(function (root, factory) {
   if (typeof exports === 'object' && typeof require === 'function') {
     module.exports = factory(require("backbone"));
   } else if (typeof define === "function" && define.amd) {
      // AMD. Register as an anonymous module.
      define(["backbone"], function(Backbone) {
        // Use global variables if the locals are undefined.
        return factory(Backbone || root.Backbone);
      });
   } else {
      // RequireJS isn't being used. Assume underscore and backbone are loaded in <script> tags
      factory(Backbone);
   }
}(this, function(Backbone) {
// A simple module to replace `Backbone.sync` with *localStorage*-based
// persistence. Models are given GUIDS, and saved into a JSON object. Simple
// as that.

// Hold reference to Underscore.js and Backbone.js in the closure in order
// to make things work even if they are removed from the global namespace

// Generate four random hex digits.
function S4() {
   return (((1+Math.random())*0x10000)|0).toString(16).substring(1);
};

// Generate a pseudo-GUID by concatenating random hexadecimal.
function guid() {
   return (S4()+S4()+"-"+S4()+"-"+S4()+"-"+S4()+"-"+S4()+S4()+S4());
};

// Our Store is represented by a single JS object in *localStorage*. Create it
// with a meaningful name, like the name you'd give a table.
// window.Store is deprectated, use Backbone.LocalStorage instead
Backbone.LocalStorage = window.Store = function(name) {
  if( !this.localStorage ) {
    throw "Backbone.localStorage: Environment does not support localStorage."
  }
  this.name = name;
  var store = this.localStorage().getItem(this.name);
  this.records = (store && store.split(",")) || [];
};

// Copy used underscore methods here
var ArrayProto = Array.prototype, ObjProto = Object.prototype, FuncProto = Function.prototype;
var nativeIndexOf      = ArrayProto.indexOf;
var slice              = ArrayProto.slice;
var nativeFilter       = ArrayProto.filter;


_extend = function(obj) {
  ps = Array.prototype.slice.call(arguments, 1);
  ps.forEach(function(source) {
    if (source) {
      for (var prop in source) {
        obj[prop] = source[prop];
      }
    }
  });
  return obj;
};

_contains = _include = function(obj, target) {
  if (obj == null) return false;
  if (nativeIndexOf && obj.indexOf === nativeIndexOf) return obj.indexOf(target) != -1;
  return any(obj, function(value) {
    return value === target;
  });
};

_chain = function(obj) {
  return _(obj).chain();
};


_filter = _select = function(obj, iterator, context) {
  var results = [];
  if (obj == null) return results;
  if (nativeFilter && obj.filter === nativeFilter) return obj.filter(iterator, context);
  each(obj, function(value, index, list) {
    if (iterator.call(context, value, index, list)) results.push(value);
  });
  return results;
};

_reject = function(obj, iterator, context) {
  return _filter(obj, function(value, index, list) {
    return !iterator.call(context, value, index, list);
  }, context);
};

//

_extend(Backbone.LocalStorage.prototype, {

  // Save the current state of the **Store** to *localStorage*.
  save: function() {
    this.localStorage().setItem(this.name, this.records.join(","));
  },

  // Add a model, giving it a (hopefully)-unique GUID, if it doesn't already
  // have an id of it's own.
  create: function(model) {
    if (!model.id) {
      model.id = guid();
      model.set(model.idAttribute, model.id);
    }
    this.localStorage().setItem(this.name+"-"+model.id, JSON.stringify(model));
    this.records.push(model.id.toString());
    this.save();
    return this.find(model);
  },

  // Update a model by replacing its copy in `this.data`.
  update: function(model) {
    this.localStorage().setItem(this.name+"-"+model.id, JSON.stringify(model));
    if (!_include(this.records, model.id.toString()))
      this.records.push(model.id.toString()); this.save();
    return this.find(model);
  },

  // Retrieve a model from `this.data` by id.
  find: function(model) {
    return this.jsonData(this.localStorage().getItem(this.name+"-"+model.id));
  },

  // Return the array of all models currently in storage.
  findAll: function() {
    // Lodash removed _#chain in v1.0.0-rc.1
    return this.records.map(function(id){
        return this.jsonData(this.localStorage().getItem(this.name+"-"+id));
      }, this)
  },

  // Delete a model from `this.data`, returning it.
  destroy: function(model) {
    if (model.isNew())
      return false
    this.localStorage().removeItem(this.name+"-"+model.id);
    this.records = _reject(this.records, function(id){
      return id === model.id.toString();
    });
    this.save();
    return model;
  },

  localStorage: function() {
    return localStorage;
  },

  // fix for "illegal access" error on Android when JSON.parse is passed null
  jsonData: function (data) {
      return data && JSON.parse(data);
  },

  // Clear localStorage for specific collection.
  _clear: function() {
    var local = this.localStorage(),
      itemRe = new RegExp("^" + this.name + "-");

    // Remove id-tracking item (e.g., "foo").
    local.removeItem(this.name);

    // Lodash removed _#chain in v1.0.0-rc.1
    // Match all data items (e.g., "foo-ID") and remove.
    var objkeys = Object.keys(local)
    var filtered = _filter(objkeys, function (k) { return itemRe.test(k); })
    filtered.forEach(function (k) { local.removeItem(k); });

    this.records.length = 0;
  },

  // Size of localStorage.
  _storageSize: function() {
    return this.localStorage().length;
  }

});

// localSync delegate to the model or collection's
// *localStorage* property, which should be an instance of `Store`.
// window.Store.sync and Backbone.localSync is deprecated, use Backbone.LocalStorage.sync instead
Backbone.LocalStorage.sync = window.Store.sync = Backbone.localSync = function(method, model, options) {
  var store = model.localStorage || model.collection.localStorage;

  var resp, errorMessage, syncDfd = Backbone.Deferred && Backbone.Deferred(); //If $ is having Deferred - use it.

  try {

    switch (method) {
      case "read":
        resp = model.id != undefined ? store.find(model) : store.findAll();
        break;
      case "create":
        resp = store.create(model);
        break;
      case "update":
        resp = store.update(model);
        break;
      case "delete":
        resp = store.destroy(model);
        break;
    }

  } catch(error) {
    if (error.code === 22 && store._storageSize() === 0)
      errorMessage = "Private browsing is unsupported";
    else
      errorMessage = error.message;
  }

  if (resp) {
    if (options && options.success) {
      if (Backbone.VERSION === "0.9.10") {
        options.success(model, resp, options);
      } else {
        options.success(resp);
      }
    }
    if (syncDfd) {
      syncDfd.resolve(resp);
    }

  } else {
    errorMessage = errorMessage ? errorMessage
                                : "Record Not Found";

    if (options && options.error)
      if (Backbone.VERSION === "0.9.10") {
        options.error(model, errorMessage, options);
      } else {
        options.error(errorMessage);
      }

    if (syncDfd)
      syncDfd.reject(errorMessage);
  }

  // add compatibility with $.ajax
  // always execute callback for success and error
  if (options && options.complete) options.complete(resp);

  return syncDfd && syncDfd.promise();
};

Backbone.ajaxSync = Backbone.sync;

Backbone.getSyncMethod = function(model) {
  if(model.localStorage || (model.collection && model.collection.localStorage)) {
    return Backbone.localSync;
  }

  return Backbone.ajaxSync;
};

// Override 'Backbone.sync' to default to localSync,
// the original 'Backbone.sync' is still available in 'Backbone.ajaxSync'
Backbone.sync = function(method, model, options) {
  return Backbone.getSyncMethod(model).apply(this, [method, model, options]);
};

return Backbone.LocalStorage;
}));